crates/jstzd/src/protocol/mod.rs

```diff
@@ -0,0 +1,166 @@
+use rust_embed::Embed;
+use serde_json::Value;
+use std::fmt::Display;
+use std::path::{Path, PathBuf};
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum ProtocolConstants {
+    Sandbox,
+    #[cfg(test)]
+    TestConstants,
+}
+
+impl Default for ProtocolConstants {
+    fn default() -> Self {
+        Self::Sandbox
+    }
+}
+
+impl Display for ProtocolConstants {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.write_str(match self {
+            Self::Sandbox => "sandbox",
+            #[cfg(test)]
+            Self::TestConstants => "test-test",
+        })
+    }
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum Protocol {
+    Alpha,
+    #[cfg(test)]
+    TestVersion,
+}
+
+impl Default for Protocol {
+    fn default() -> Self {
+        Self::Alpha
+    }
+}
+
+impl Protocol {
+    fn hash(self) -> &'static str {
+        match self {
+            Protocol::Alpha => "ProtoALphaALphaALphaALphaALphaALphaALphaALphaDdp3zK",
+            #[cfg(test)]
+            Protocol::TestVersion => "test-hash",
+        }
+    }
+
+    fn parameter_file(self, constants: ProtocolConstants) -> PathBuf {
+        Path::new(&constants.to_string()).join(self.hash())
+    }
+}
+
+#[cfg(not(test))]
+#[derive(Embed)]
+#[folder = "$CARGO_MANIFEST_DIR/resources/protocol_parameters/"]
+pub struct ProtocolParameterFile;
+
+#[cfg(test)]
+#[derive(Embed)]
+#[folder = "$CARGO_MANIFEST_DIR/resources/test/protocol_parameters/"]
+pub struct ProtocolParameterFile;
+
+#[derive(Default)]
+pub struct ProtocolParameterBuilder {
+    protocol: Protocol,
+    constants: ProtocolConstants,
+    path: Option<PathBuf>,
+}
+
+impl ProtocolParameterBuilder {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn set_protocol(&mut self, protocol: Protocol) -> &mut Self {
+        self.protocol = protocol;
+        self
+    }
+
+    pub fn set_constants(&mut self, constants: ProtocolConstants) -> &mut Self {
+        self.constants = constants;
+        self
+    }
+
+    pub fn set_path(&mut self, path: &str) -> &mut Self {
+        self.path = Some(PathBuf::from(path));
+        self
+    }
+
+    pub async fn build(self) -> anyhow::Result<PathBuf> {
+        let f = ProtocolParameterFile::get(
+            self.protocol
+                .parameter_file(self.constants)
+                .to_str()
+                .unwrap(),
+        )
+        .unwrap();
```

Comment:
We probably want `?` here

Reply:
I've replaced unnecessary `unwrap`s with `?`.